### PR TITLE
Dependencies and template upgrade

### PR DESCRIPTION
**Pull request by ashishawasthi**
Updates and refactoring done:
- updated dependencies
- code changes for the dependency updates
- took updates from https://github.com/flutter/samples/tree/main/game_template

### CODE REVIEW

File: lib/main.dart [hunk 1]
```diff
@@ -116,7 +117,7 @@ class MyApp extends StatelessWidget {
           routes: [
             GoRoute(
                 path: 'play',
-                pageBuilder: (context, state) => buildTransition(
+                pageBuilder: (context, state) => buildTransition<void>(
```
Comment (filiph) [inline lib/main.dart:120]: Is this needed now?

File: lib/src/audio/sounds.dart [hunk 2]
```diff
@@ -1,13 +1,13 @@
 List<String> soundTypeToFilename(SfxType type) {
   switch (type) {
-    case SfxType.drawX:
-      return [
+    case SfxType.huhsh:
```
Comment (filiph) [inline lib/src/audio/sounds.dart:3]: Why this change? Just curious.

File: lib/src/in_app_purchase/in_app_purchase.dart [hunk 3]
```diff
@@ -85,19 +102,20 @@ class InAppPurchaseController extends ChangeNotifier {
...
   /// Subscribes to the [inAppPurchaseInstance.purchaseStream].
   void subscribe() {
+    _subscription?.cancel();
     _subscription =
         inAppPurchaseInstance.purchaseStream.listen((purchaseDetailsList) {
       _listenToPurchaseUpdated(purchaseDetailsList);
     }, onDone: () {
       _subscription?.cancel();
-    }, onError: (error) {
+    }, onError: (dynamic error) {
```
Comment (filiph) [inline lib/src/in_app_purchase/in_app_purchase.dart]: Why this change?

File: pubspec.yaml [hunk 4]
```diff
@@ -7,29 +7,30 @@ publish_to: 'none'
 version: 1.7.8+35
 
 environment:
-  sdk: ">=2.17.0 <3.0.0"
+  sdk: ">=2.17.0-0 <3.0.0"
```
Comment (filiph) [inline pubspec.yaml:10]: Why this change?

**Review by filiph — COMMENTED** [review]
This looks good! I'm just confused about one or two changes, that's all.

File: lib/main.dart (hunk 1 above)
Comment (ashishawasthi) [reply lib/main.dart:120]: Assuming the purpose of this repo is to show how to use game_template, an explicit <void> may help developers understand that a custom type can be used in Page<T>. This can be taken out though, as it is unnecessary code for this repo

File: lib/src/audio/sounds.dart (hunk 2 above)
Comment (ashishawasthi) [reply lib/src/audio/sounds.dart:3]: Since the sound can be used for a purpose other than drawing X

File: pubspec.yaml (hunk 4 above)
Comment (ashishawasthi) [reply pubspec.yaml:10]: I believe this excludes prerelease versions, but I'm not sure about this. This is taken from game_template repo, while trying to reduce differences between the repos, so that developers trying to understand the changes required over game_template, dont have to look at the differences/changes they don't need

File: lib/src/in_app_purchase/in_app_purchase.dart (hunk 3 above)
Comment (ashishawasthi) [reply lib/src/in_app_purchase/in_app_purchase.dart]: this change should not be there, I'll revert

**Comment by filiph** [comment]
Makes sense! Thanks for the clarification!
